domino.go

```diff
@@ -1073,48 +1217,69 @@ func (d *Scan) Build() *dynamodb.ScanInput {
  ** 		   the returned channel.
  **
  */
-func (d *Scan) ExecuteWith(ctx context.Context, dynamodb DynamoDBIFace, nextItem interface{}, opts ...request.Option) (c chan interface{}, e chan error) {
+func (d *ScanInput) ExecuteWith(ctx context.Context, db DynamoDBIFace, opts ...request.Option) (out *ScanOutput) {
 
-	c = make(chan interface{})
-	e = make(chan error)
+	out = &ScanOutput{
+		ctx:   ctx,
+		limit: d.Limit,
+	}
 
-	go func() {
-		defer close(c)
-		defer close(e)
+	q := d.Build()
 
-		var count int64
-	Execute:
-		if d.Limit != nil && count >= *d.Limit {
+	out.outputFunc = func() (o *dynamodb.ScanOutput, err error) {
+		if q == nil {
 			return
 		}
-		out, err := dynamodb.ScanWithContext(ctx, d.Build(), opts...)
+		o, err = db.ScanWithContext(ctx, q, opts...)
 		if err != nil {
-			e <- handleAwsErr(err)
+			err = handleAwsErr(err)
 			return
 		}
 
-		for _, item := range out.Items {
-			err = dynamodbattribute.UnmarshalMap(item, &nextItem)
+		if o.LastEvaluatedKey != nil {
+			q.ExclusiveStartKey = o.LastEvaluatedKey
+		} else {
+			q = nil
+		}
+		return
+	}
+
+	return
 
-			if err != nil {
-				e <- handleAwsErr(err)
+}
+
+func (o *ScanOutput) Results(next func() interface{}) (err error) {
+	err = o.Error
+	if err != nil || o.outputFunc == nil {
+		return
+	}
+	var count int64
+	for {
```

Comment:
This one is a bit mind-bending -- wondering if there is an off-by-one in the o.limit check below -- not sure how page boundaries work, and if `out.Items` is empty will we even do the check since the range won't run?